lib/modules/manager/npm/extract/pnpm.ts

```diff
@@ -135,3 +142,150 @@ export async function detectPnpmWorkspaces(
+      logger.trace({
+        packagePath,
+        packageName,
+        version,
+      });
+      lockedVersions[packageName] = version;
+    }
+    return {
+      lockedVersions,
+      lockfileVersion,
+    };
+  } catch (err) {
+    logger.debug({ filePath, err }, 'Warning: Exception parsing pnpm lockfile');
+    return { lockedVersions: {} };
+  }
+}
+
+export function getConstraints(
+  lockfileVersion: number,
+  constraints?: string
+): string {
+  let newConstraints = constraints;
+
+  // find matching lockfileVersion and use its constraints
+  // if no match found use lockfileVersion 5
+  // lockfileVersion 5 is the minimum version required to generate the pnpm-lock.yaml file
+  const { lowerBound, upperBound, lowerConstraint, upperConstraint } =
+    lockToPnpmVersionMapping.find(
+      (m) => m.lockfileVersion === lockfileVersion
+    ) ?? {
+      lockfileVersion: 5.0,
+      lowerBound: '2.24.0',
+      upperBound: '3.5.0',
+      lowerConstraint: '>=3',
+      upperConstraint: '<3.5.0',
+    };
+
+  // inorder to ensure that the constraint doesn't allow any pnpm versions that can't generate the extracted lockfileVersion
+  // compare the current constraint to the lowerBound and upperBound of the lockfileVersion
+  // if the current constraint is not comaptible, add the lowerConstraint and upperConstraint, whichever is needed
+  if (newConstraints) {
+    // if constraint satisfies versions lower than lowerBound add the lowerConstraint to narrow the range
+    if (semver.satisfies(lowerBound, newConstraints)) {
+      newConstraints += ` ${lowerConstraint}`;
+    }
+
+    // if constraint satisfies versions higher than upperBound add the upperConstraint to narrow the range
+    if (
+      upperBound &&
+      upperConstraint &&
+      semver.satisfies(upperBound, newConstraints)
+    ) {
+      newConstraints += ` ${upperConstraint}`;
+    }
+  }
+  // if no constraint is present, add the lowerConstraint and upperConstraint corresponding to the lockfileVersion
+  else {
+    newConstraints = `${lowerConstraint}${
+      upperConstraint ? ` ${upperConstraint}` : ''
+    }`;
+  }
+
+  return newConstraints;
+}
+
+/**
+ pnpm lockfiles have corresponding version numbers called "lockfileVersion"
+ each lockfileVersion can only be generated by a certain pnpm version ranges
+ eg. lockfileVersion: 5.4 can only be generated by pnpm version >=7 && <8
+ official list can be found here : https:github.com/pnpm/spec/tree/master/lockfile
+ we use the mapping present below to find the compatible pnpm version range for a given lockfileVersion
+
+ the various terms used in the mapping are explained below:
+ lowerConstriant : lowest pnpm version that can generate the lockfileVersion
+ upperConstraint : highest pnpm version that can generate the lockfileVersion
+ lowerBound      : highest pnpm version that is less than the lowerConstraint
+ upperBound      : lowest pnpm version that is greater than upperConstraint
```

Comment:
Why do you need e.g. lowerBound if you already have lowerConstraint?

Reply:
lowerBound is used to check if lowerConstraint needs to be added to existing constraint or not.

For eg. if `lockfileVersion = 6.0` and existing constraint is `>7`. lowerBound(7.32.0) is in the constraint's range but version `7.32.0`  isn't compatible with lockfileVersion `6.0`, so we need to add lowerConstraint ( >=8 ). However if the existing constraint was `>8` then we wouldn't need to add the lowerConstraint.

We can add both lower and upperConstraints without cheking first, but that would make the constraint string bigger for no good reason. Also we follow the same logic for `npm` and `yarn` constraints.